modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/IndexMinMaxRule.java

```diff
@@ -0,0 +1,95 @@
+import org.apache.ignite.internal.processors.query.calcite.rel.AbstractIndexScan;
+import org.apache.ignite.internal.processors.query.calcite.rel.IgniteAggregate;
+import org.apache.ignite.internal.processors.query.calcite.rel.IgniteIndexBound;
+import org.apache.ignite.internal.processors.query.calcite.rel.IgniteIndexScan;
+import org.apache.ignite.internal.processors.query.calcite.schema.IgniteIndex;
+import org.apache.ignite.internal.processors.query.calcite.schema.IgniteTable;
+import org.apache.ignite.internal.processors.query.calcite.trait.RewindabilityTrait;
+import org.apache.ignite.internal.util.typedef.F;
+import org.immutables.value.Value;
+
+/**
+ * Tries to optimize MIN() and MAX() so that taking only first or last index record is engaged.
+ */
+@Value.Enclosing
+public class IndexMinMaxRule extends RelRule<IndexMinMaxRule.Config> {
+    /** */
+    public static final IndexMinMaxRule INSTANCE = Config.DEFAULT.toRule();
+
+    /** Ctor. */
+    private IndexMinMaxRule(IndexMinMaxRule.Config cfg) {
+        super(cfg);
+    }
+
+    /** */
+    @Override public void onMatch(RelOptRuleCall call) {
+        IgniteAggregate aggr = call.rel(0);
+        IgniteIndexScan idxScan = call.rel(1);
+        IgniteTable table = idxScan.getTable().unwrap(IgniteTable.class);
+        IgniteIndex idx = table.getIndex(idxScan.indexName());
+
+        if (
+            table.isIndexRebuildInProgress() ||
+                idxScan.condition() != null ||
+                aggr.getGroupCount() > 0 ||
+                aggr.getAggCallList().stream().filter(a -> a.getAggregation().getKind() == SqlKind.MIN
+                    || a.getAggregation().getKind() == SqlKind.MAX).count() != 1 ||
+                idx.collation().getFieldCollations().isEmpty() ||
+                idx.collation().getFieldCollations().get(0).getFieldIndex() != idxScan.requiredColumns().asList().get(0)
+        )
+            return;
+
+        SqlAggFunction aggFun = aggr.getAggCallList().get(0).getAggregation();
+        boolean firstIdxValue = (aggFun.getKind() == SqlKind.MIN) !=
+            idx.collation().getFieldCollations().get(0).getDirection().isDescending();
+
+        IgniteIndexBound newAggrInput = new IgniteIndexBound(
+            idxScan.getTable(),
+            idxScan.getCluster(),
+            idxScan.getTraitSet().replace(RewindabilityTrait.REWINDABLE),
+            idxScan.indexName(),
+            firstIdxValue,
+            idx.collation()
+        );
+
+        call.transformTo(aggr.clone(aggr.getCluster(), F.asList(newAggrInput)));
+    }
+
+    /** The rule config. */
+    @Value.Immutable
+    public interface Config extends RelRule.Config {
+        /** */
+        IndexMinMaxRule.Config DEFAULT = ImmutableIndexMinMaxRule.Config.of()
+            .withDescription("IndexMinMaxRule")
+            .withOperandSupplier(r -> r.operand(IgniteAggregate.class)
+                .oneInput(i -> i.operand(AbstractIndexScan.class).anyInputs()));
```

Comment:
Why `AbstractIndexScan` but not `IgniteIndexScan` (physical node)? Using only physical node will reduce search space in my opinion.

Reply:
Fixed